src/Products/PluginIndexes/CompositeIndex/CompositeIndex.py

```diff
@@ -0,0 +1,599 @@
+            except TypeError:
+                return 0
+        else:
+            # we have an existing entry for this document, and we need
+            # to figure out if any of the keywords have actually changed
+            if type(oldKeywords) is not OOSet:
+                oldKeywords = OOSet(oldKeywords)
+            newKeywords = OOSet(newKeywords)
+            fdiff = difference(oldKeywords, newKeywords)
+            rdiff = difference(newKeywords, oldKeywords)
+            if fdiff or rdiff:
+                # if we've got forward or reverse changes
+                self._unindex[documentId] = list(newKeywords)
+                if fdiff:
+                    self.unindex_objectKeywords(documentId, fdiff)
+                if rdiff:
+                    for kw in rdiff:
+                        self.insertForwardIndexEntry(kw, documentId)
+
+        return 1
+
+    def unindex_objectKeywords(self, documentId, keywords):
+        """ carefully unindex the object with integer id 'documentId'"""
+
+        if keywords is not None:
+            for kw in keywords:
+                self.removeForwardIndexEntry(kw, documentId)
+
+    def unindex_object(self, documentId):
+        """ carefully unindex the object with integer id 'documentId'"""
+
+        keywords = self._unindex.get(documentId, None)
+        self.unindex_objectKeywords(documentId, keywords)
+        try:
+            del self._unindex[documentId]
+        except KeyError:
+            LOG.debug('%s: Attempt to unindex nonexistent '
+                      'document with id %s' %
+                      (self.__class__.__name__, documentId),
+                      error=sys.exc_info())
+
+    def _get_object_keywords(self, obj):
+        """ returns permutation list of object keywords """
+
+        components = self.getIndexComponents()
+        kw_list = []
+
+        for c in components:
+            kw = self._get_component_keywords(obj, c)
+            # skip if keyword list is empty
+            if not kw:
+                continue
+            kw = tuple([(c.id, k) for k in kw])
+            kw_list.append(kw)
+
+        # permute keyword list in order to support any combination and
+        # number (n > 1) of components in query
+        pkl = []
+        c_list = product(*kw_list)
+
+        for c in c_list:
+            for r in range(MIN_COMPONENTS, len(c) + 1):
+                p = combinations(c, r)
+                pkl.extend(p)
+
+        return tuple(pkl)
+
+    def _get_component_keywords(self, obj, component):
+
+        if component.meta_type == 'FieldIndex':
+            attr = component.attributes[-1]
+            try:
+                datum = getattr(obj, attr)
+                if safe_callable(datum):
+                    datum = datum()
+            except (AttributeError, TypeError):
+                datum = _marker
+            if isinstance(datum, list):
+                datum = tuple(datum)
+            return (datum,)
```

Comment:
Could you shorten this, by using UnIndex' _get_object_datum method? For example:
```
if component.meta_type == 'FieldIndex':
    attr = component.attributes[-1]
    datum = self._get_object_datum(obj, attr)
    if isinstance(datum, list):
        datum = tuple(datum
    return (datum,)
```